src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/FileMoveNotificationListener.cs

```diff
@@ -56,190 +58,141 @@ internal class FileMoveNotificationListener : IFileMoveNotificationListener
+                string[] destinationFolders = destinationFolderPath.Split(Delimiter.Path, StringSplitOptions.RemoveEmptyEntries);
 
-            void RecursiveTryGetFilesToMove(IFileMoveItem? item, ref List<(string file, string destination)>? filesToMove)
-            {
-                if (item is null)
+                // Since this rename only moves the location of the file to another directory, it will use the SyncNamespaceDocumentAction in Roslyn as the rename action within this set.
+                // The logic for selecting this rename action can be found here: https://github.com/dotnet/roslyn/blob/960f375f4825a189937d4bfd9fea8162ecc63177/src/Workspaces/Core/Portable/Rename/Renamer.cs#L133-L136
+                RenameDocumentActionSet renameActionSet = await RenameDocumentAsync(currentDocument, s_renameOptions, null, destinationFolders);
+                if (renameActionSet.ApplicableActions.IsEmpty || renameActionSet.ApplicableActions.Any(aa => aa.GetErrors().Any()))
                 {
-                    return;
+                    continue;
                 }
 
-                if (item.IsFolder)
-                {
-                    if (item is not ICopyPasteItem copyPasteItem)
-                    {
-                        return;
-                    }
-
-                    foreach (IFileMoveItem child in copyPasteItem.Children.Cast<IFileMoveItem>())
-                    {
-                        RecursiveTryGetFilesToMove(child, ref filesToMove);
-                    }
-                }
-                else
-                {
-                    bool isCompileItem = StringComparers.ItemTypes.Equals(item.ItemType, Compile.SchemaName);
+                // Getting the rename message requires an instance of RenameDocumentAction.
+                // We only need to set this message text once for the lifetime of the class, since it isn't dynamic.
+                // Even though it isn't dynamic, it does get localized appropriately in Roslyn.
+                // The text in English is "Sync namespace to folder structure".
+                _renameMessage ??= renameActionSet.ApplicableActions.First().GetDescription();
 
-                    if (item.WithinProject && isCompileItem && !item.IsLinked && !item.IsFolder)
-                    {
-                        filesToMove ??= new();
-                        filesToMove.Add((item.Source, item.Destination));
-                    }
-                }
+                _renameActionSets.Add(itemToMove.Source, renameActionSet);
             }
 
-            async Task<List<(Renamer.RenameDocumentActionSet, string)>> GetNamespaceUpdateActionsAsync()
-            {
-                List<(Renamer.RenameDocumentActionSet, string)> actions = new();
+            return;
 
-                foreach ((string filenameWithPath, string destination) in filesToMove)
+            static IEnumerable<IFileMoveItem> GetFilesToMoveRecursive(IEnumerable<IFileMoveItem> items)
```

Comment:
Simplified the recursive logic by making it a single local function. It takes in the items and then if the item is a folder, calls this same function on each of its children. The result is getting every child non-folder hierarchically from a set of items.

Reply:
In general, it's good to avoid stack recursion, which can overflow. This can be rewritten with a `Queue<IFileMoveItem>` containing initially the first items, followed by _while not empty, pop item and process it, if it contains more items, stick them in the queue, end while_.

Reply:
I've rewritten it locally to use a `Queue`. The reason I didn't do this initially is that using a `Queue` changes the order in which files would be processed. Thinking about it a bit more, since this order only impacts the namespace update process, it wouldn't be something that should affect the user any differently. Either all namespace updates are applied or none of them. The only effect in ordering is the way Roslyn would process the changes. I have no idea if that makes a difference, though.

I'll test it a bit before I push the commit.

Reply:
I tested it and saw no ill-effects from using the queue over recursion. I've committed these changes.